### DISCUSSION TITLE
Alpha bit while saving in PNG

### DISCUSSION THREAD

**Discussion by 2246sam20** [ImageMagick/ImageMagick#7546] [Help] [answered]
### ImageMagick version

 7.1.1-37

### Operating system, version and so on

Windows 11

### Description

Hello! I am trying to save image object in PNG format and I have set the depth to 16 bits, So every channel(RGBA) should have a size of 16 bits. But when I save the image I am getting the size of alpha bit to be one.
 I am checking the image info using the identify command.
Below is the output of the identify command.

##########################################################################
  Image:
    Filename: output.png
    Permissions: rw-rw-rw-
    Format: PNG (Portable Network Graphics)
    Mime type: image/png
    Class: DirectClass
    Geometry: 1080x1792+0+0
    Units: Undefined
    Colorspace: sRGB
    Type: TrueColorAlpha
    Endianness: Undefined
    Depth: 16-bit
    Channels: 4.0
    Channel depth:
      Red: 16-bit
      Green: 16-bit
      Blue: 16-bit
      Alpha: 1-bit
##########################################################################

The code snippet: 
```
	int width = std::stoi(page.width) == 0 ? CONSTANTS::MAX_WIDTH : std::stoi(page.width);
	int height = std::stoi(page.height) == 0 ? CONSTANTS::MAX_HEIGHT : std::stoi(page.height);
	int rot = 90;
	Magick::Image pageImage(Magick::Geometry(width, height), Magick::Color("transparent"));
	pageImage.alpha(true);
	
	 //some operations on the image object like rendering diffferent shapes.
	 
	pageImage.rotate(rot);
	std::string new_width = std::to_string(CONSTANTS::TIZEN_IWB_WIDTH);
	std::string new_height = std::to_string(CONSTANTS::TIZEN_IWB_HEIGHT);
	std::string geomString = new_width + "x" + new_height + '!'; // to dicard aspect ratio
	
	pageImage.resize(Magick::Geometry(geomString));
	pageImage.magick("PNG");
	pageImage.depth(CONSTANTS::BIT_DEPTH);
	pageImage.write(pageOutputFilePath);
```

Any help will be appreciated.

**Comment by snibgo** [comment]
The reported `Channel depth:` is after IM has read the image into memory.

If you want to save a PNG as 4 channels, with 16 bits per channel = 4\*16 = 64 bits. Save the image as `PNG64`.

EDIT: Corrected embarrassingly incorrect calculation.

**Reply by 2246sam20** [reply]
Lets say I want to save 8 bits for each channel..
So this will not work `pageImage.depth(8).` ?

I tried with the following methods too:
```
1. pageImage.write("png32:" + pageOutputFilePath);
2. pageImage.defineValue("png", "bit-depth", "8");
```

But none of them works. Sometime Magick  makes red channel's depth to 1 bit and  sometimes alpha to 1 bit, I think it is doing some kind of optimization. 
What can be the workaround here?
P.S- I have updated the code snippet above, it was incorrect ..my bad :/
Thanks in Advance.

**Comment by snibgo** [comment, marked as answer]
IM doesn't tell you how many bits are stored per channel *in the file*.

When IM says this:
```
Alpha: 1-bit
```
... it means that after reading the file into memory, the alpha channel needs 1 bit to encode it. Thus, the alpha channel has two values or fewer.

If your image has only two values in the alpha channel, you might save the image to a file with 1 bit per alpha value, or 2 bits, or 16 bits or any number of bits, but there will be only two values, and when IM reads that file and reports information, it will still say `Alpha: 1-bit`.

Why do you care about how many bits are used in the file?

**Reply by 2246sam20** [reply]
Thanks for the explanation.
Actually I am saving a magick::image object after some operations to a png file.
And I have some other cpp module which tries to load this png file but its throwing some error while loading png buffer data.
